src/lib/html_renderer.js

```diff
@@ -0,0 +1,42 @@
+const SPAN_CLOSE = '</span>';
+
+import {escapeHTML} from './utils';
+
+export default class HTMLRenderer {
+  constructor(tree, options) {
+    this.buffer = "";
+    this.classPrefix = options.classPrefix;
+    tree.walk(this);
+  }
+
+  // renderer API
+
+  addText(text) {
+    this.buffer += escapeHTML(text)
+  }
+
+  openNode(node) {
+    let className = node.kind;
+    if (!node.kind) return;
```

Comment:
What's the meaning of nodes with no `kind`? It looks strange that the renderer has to check it all the time.

Reply:
It's for unnamed modes... ie, no `className`... there really isn't a reason to have a special node in the tree for them since they don't affect the output one bit.  You could argue this is a premature optimization except storing the tree itself has measurable memory implications relative to just using a string.  So I didn't see the need to track pointless nodes.

Technically if we were going to store `relevance` in the tree (which we don't now) then we'd have to revisit this as we'd actually need the node then.

---

You might have a point though that this should be a responsibility of the parser though, not the tree itself, that now the tree knows too much about the parser internals.  I'll give that some thought.

Reply:
**Oh, wait this is the HTML renderer.**  It's POSSIBLE that the tree might include notes that don't need require any output (such as future `relevance` nodes mentioned above) so hence this check is done on each node.  

It's not the tree's responsibility to know what should and should not be output, that's the renderers job, hence the check here.

Reply:
Hm. Maybe `kind` is not the best choice than... It's not that intuitive that `kind` reflects the need to output something when walking the tree.
Bwt, shouldn't relevance be an attribute of an existing node (not a separate node)?

Reply:
>  It's not that intuitive that kind reflects the need to output something when walking the tree.

**You have to think about them separately.**  Kind ONLY tells the tree what kind of node it is.  It has no other meaning INSIDE the class.

But I can make a helpful function to wrap this up conceptually...

> Bwt, shouldn't relevance be an attribute of an existing node (not a separate node)?

I'm not sure I follow... if we added relevance it would be an attribute of a node.  A node has a given relevance...

Reply:
I think this help a lot:

https://github.com/highlightjs/highlight.js/pull/2404/commits/4d3078325c33155b8fba7495a9480127fe438422

Reply:
> if we added relevance it would be an attribute of a node. A node has a given relevance...

Right. Then I'm not sure I understand this:

> Technically if we were going to store relevance in the tree (which we don't now) then we'd have to revisit this as we'd actually need the node then.

(Not that it is important.)

Reply:
I mean that right now the parser skips emitting nodes that don't have a className... because they don't affect output.  BUT if we wanted to store relevance in the tree then we'd need nodes to store it.

Right now if you had a bunch of rules but no classNames you'd end up with ONE large test node... with no where to store relevance...

I suppose another way would be to just accumulate relevance into whatever the current node is, but that would make it harder for 3rd parties to work with relevance data.

Reply:
I may still rip out this whole thing before 10.0, LOL - leaving it only as a reference implementation.  It concerns me that there are increased memory requirements.  I don't think anyone using this on their blog is going to notice but someone using this on a backend system to process thousands of snippets or some sort of more "industrial" use this new version might have significantly different requirements than the old version.

So I weigh that against: is adding a parse tree that only a tiny portion of people have actually requested worth it?

Making it optional is another choice, we'd just have to alter the test suite to run both both emitters... that starts to sound like a lot of headache though.

So I think what might happen is the emitter API stays (that part of the work is a big improvement) but the parse tree implementation because reference and we allow people to configure their own emitters.  I really think this type of thing should be easy to do (emit to a tree instead of string), but that doesn't mean we  have to ship it by default.

Then by default we'd ship the "FastHTMLRender" emitter. :-)

Reply:
So I take back everything I said here if the parse tree is actually FASTER and uses less memory then it's a clear win win, though I'm still puzzled.

Reply:
I got similar figures... but I'm not sure if `time` counts child processes?